### PR TITLE
Add fuzz package for running go-fuzz

### LINKED ISSUE #31: Add more fuzzing-like tests
e.g. ideally, stuff like https://github.com/lazyledger/smt/issues/25 would be found by this kind of testing

### PR DESCRIPTION
**Pull request by cuonglm**
While at it, also add oss fuzz build script for to prepare for adding to
oss-fuzz later.

Fixes #31

### CODE REVIEW

**Comment by cuonglm** [comment]
cc @odeke-em

**Review by odeke-em — APPROVED** [review]
LGTM, thank you @cuonglm!

**Review by liamsi — COMMENTED** [review]
Just to confirm my understanding: 

- The action `.github/workflows/fuzz_build.yml` is necessary to have a build which can then be included in oss-fuzz? And same with `oss-fuzz-build.sh`?
- the code under `fuzz/fuzz.go` is some very basic fuzzing to see if we panic on any random input/operation

Is that right?

Looks all good to me. Thanks so much!

File: oss-fuzz-build.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash -eu
```
Comment (adlerjohn) [inline oss-fuzz-build.sh:1]: What is this file used for?

File: .github/workflows/fuzz_build.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    branches:
+      - master
+
+jobs:
+  build_fuzz:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-go@v2
+        with:
+          go-version: "1.16"
```
Comment (adlerjohn) [inline .github/workflows/fuzz_build.yml:15]: 1.15 is used in the other CI workflow, so use it here too.

File: oss-fuzz-build.sh (hunk 1 above)
Comment (cuonglm) [reply oss-fuzz-build.sh:1]: This will be used by oss-fuzz to build the `fuzz` package when submit this repo to it.

**Comment by cuonglm** [comment]
@liamsi 

> The action .github/workflows/fuzz_build.yml is necessary to have a build which can then be included in oss-fuzz? And same with oss-fuzz-build.sh?

No, `.github/workflows/fuzz_build.yml` is just our local testing to make sure the fuzz package can be built. When we submit this repo to oss-fuzz, we will use the `oss-fuzz-build.sh` script to built the necessary object file for lib fuzzer to work. The idea is that we just submit a PR to oss-fuzz once, then continuously adding fuzzer in our repo, oss-fuzz will checkout the latest and run the fuzzing daily.

See discussion in https://github.com/google/oss-fuzz/pull/5631#issuecomment-822532399 for more details

> the code under fuzz/fuzz.go is some very basic fuzzing to see if we panic on any random input/operation

Yes, exactly.

File: .github/workflows/fuzz_build.yml (hunk 2 above)
Comment (cuonglm) [reply .github/workflows/fuzz_build.yml:15]: Done

**Review by liamsi — APPROVED** [review]
Thanks for the clarification!

LGTM

File: oss-fuzz-build.sh (hunk 1 above)
Comment (cuonglm) [reply oss-fuzz-build.sh:1]: @adlerjohn See my comment https://github.com/celestiaorg/smt/pull/51#issuecomment-886845287 for more details.

**Comment by cuonglm** [comment]
@liamsi @adlerjohn which email I should use as primary contact for celestiaorg/smt?

See https://google.github.io/oss-fuzz/getting-started/new-project-guide/#primary for what primary contact mean.

**Comment by liamsi** [comment]
@cuonglm, @musalbas will setup an account / list dev@celestia.org that will notify all relevant people.

**Comment by odeke-em** [comment]
> @cuonglm, @musalbas will setup an account / list dev@celestia.org that will notify all relevant people.

@liamsi I'd highly recommend separating out to say security@celestia.org lest the dev list get drowned out and folks shall get lots of other messages coming to dev :-)

**Comment by cuonglm** [comment]
@liamsi Thanks! But to be able to see the dashboard, you need an individual Google Account, according to https://google.github.io/oss-fuzz/getting-started/new-project-guide/#primary

So please list any email you want to be included. I'm currently listing only people from Orijtech https://github.com/google/oss-fuzz/pull/6134/files#diff-93b109d8f5b36384edc36042a809c0afdb84d4361312b1fb0f42cb46d66648e5R2

**Comment by liamsi** [comment]
OK, for now please just use my email: ismail@celestia.org

I guess it's easy to submit a PR oss-fuzz to change that later in case we create a dedicated account for this (security@celestia.org is also a list not an individual account afaik).

**Comment by cuonglm** [comment]
> OK, for now please just use my email: [ismail@celestia.org](mailto:ismail@celestia.org)
> 
> I guess it's easy to submit a PR oss-fuzz to change that later in case we create a dedicated account for this ([security@celestia.org](mailto:security@celestia.org) is also a list not an individual account afaik).

Done.

**Pull request merged**